src/Castle.Windsor/MicroKernel/Arguments.cs

```diff
@@ -19,209 +19,204 @@ namespace Castle.MicroKernel
 	using System.Collections.Generic;
 
 	using Castle.Core;
-	using Castle.MicroKernel.Context;
+	using Castle.Windsor;
 
 	/// <summary>
-	///   Represents collection of arguments used when resolving a component.
+	/// Represents a collection of named and typed arguments used for dependencies resolved via <see cref="IWindsorContainer.Resolve{T}(Castle.MicroKernel.Arguments)"/>
+	/// See: https://github.com/castleproject/Windsor/blob/master/docs/arguments.md
 	/// </summary>
-	public class Arguments : IDictionary
+	public sealed class Arguments
+		: IEnumerable<KeyValuePair<object, object>> // Required for collection initializers
 	{
-		protected IDictionary arguments;
+		private static readonly ArgumentsComparer Comparer = new ArgumentsComparer();
 
-		public Arguments(object namedArgumentsAsAnonymousType, params IArgumentsComparer[] customComparers)
-			: this(new ReflectionBasedDictionaryAdapter(namedArgumentsAsAnonymousType), customComparers)
+		private readonly Dictionary<object, object> dictionary;
+
+		/// <summary>
+		/// Initializes a new instance of the <see cref="Arguments"/> class that is empty.
+		/// </summary>
+		public Arguments()
 		{
+			dictionary = new Dictionary<object, object>(Comparer);
 		}
 
-		public Arguments(IDictionary values, params IArgumentsComparer[] customComparers)
-			: this(customComparers)
+		/// <summary>
+		/// Initializes a new instance of the <see cref="Arguments"/> class that contains elements copied from the specified <see cref="Arguments"/>.
+		/// </summary>
+		public Arguments(Arguments arguments)
 		{
-			foreach (DictionaryEntry entry in values)
-			{
-				Add(entry.Key, entry.Value);
-			}
+			dictionary = new Dictionary<object, object>(arguments.dictionary, Comparer);
 		}
 
-		public Arguments(object[] typedArguments, params IArgumentsComparer[] customComparers)
-			: this(customComparers)
+		IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+		public IEnumerator<KeyValuePair<object, object>> GetEnumerator() => dictionary.GetEnumerator();
+
+		public void Add(object key, object value) => dictionary.Add(key, value);
+		public bool Contains(object key) => dictionary.ContainsKey(key);
+		public int Count => dictionary.Count;
+		public void Remove(object key) => dictionary.Remove(key);
+
+		public object this[object key]
 		{
-			foreach (var @object in typedArguments)
+			get
 			{
-				if (@object == null)
+				if (dictionary.TryGetValue(key, out object value))
 				{
-					throw new ArgumentNullException("typedArguments",
-					                                "Given array has null values. Only non-null values can be used as arguments.");
+					return value;
 				}
-
-				Add(@object.GetType(), @object);
+				return null;
```

Comment:
Yes, adhering to the `IDictionary` behaviour because there is heaps of Windsor code that expects null to be returned and then it'll create something and set it. Those things could use `Contains` but that is just how it is for now and I didn't want to make another large change at the moment.

Reply:
It's a breaking change to start throwing instead of returning null, so whatever you first ship is probably how it should stay.